src/shell/pipe.rs

```diff
@@ -159,15 +187,30 @@ impl<'a> PipelineExecution for Shell<'a> {
             .drain(..).map(|mut job| {
                 (job.build_command(), job.kind)
             }).collect();
-
-        if let Some(ref stdin) = pipeline.stdin {
-            if let Some(command) = piped_commands.first_mut() {
-                match File::open(&stdin.file) {
-                    Ok(file) => unsafe { command.0.stdin(Stdio::from_raw_fd(file.into_raw_fd())); },
-                    Err(err) => {
-                        let stderr = io::stderr();
-                        let mut stderr = stderr.lock();
-                        let _ = writeln!(stderr, "ion: failed to redirect stdin into {}: {}", stdin.file, err);
+        match pipeline.stdin {
+            None => (),
+            Some(Input::File(ref filename)) => {
+                if let Some(command) = piped_commands.first_mut() {
+                    match File::open(filename) {
+                        Ok(file) => unsafe {
+                            command.0.stdin(Stdio::from_raw_fd(file.into_raw_fd()));
+                        },
+                        Err(e) => {
+                            eprintln!("ion: failed to redirect '{}' into stdin: {}", filename, e);
+                        }
+                    }
+                }
+            },
+            Some(Input::HereString(ref string)) => {
+                if let Some(command) = piped_commands.first_mut() {
+                    match unsafe { crossplat::stdin_of(string) } {
```

Comment:
Adding the following fixes the issue with the "foo" argument:

```rust
let result = if string.ends_with('\n') {
    unsafe { crossplat::stdin_of(string) }
} else {
    let string = [string, "\n"].concat();
    unsafe { crossplat::stdin_of(&string) }
};
```

Doing a match on that `result` will get you what you want.